src/Git/Index/Update.hs

```diff
@@ -60,6 +69,41 @@ updateIndex relpath = do
       Normal ie -> Just ie
       _ -> Nothing
 
+parseContent :: MonadFail m => Parser a -> Path.AbsFile -> IO (m a)
+parseContent parser path = do
+    content <- openBinaryFile path ReadMode >>= LBS.hGetContents
+    pure $ lazyParseOnly (parser <* endOfInput) content
+
+getIgnores :: (MonadIO m, MonadError GitError m) => Path.AbsDir -> m Ignores
+getIgnores dirP = do
+    eMfIgnores <- liftIO $ (
+        try $ parseContent ignoresP $ dirP </> Path.relFile ".gitignore"
+        :: IO (Either IOException (Either String Ignores)))
+    case eMfIgnores of
+        Left _ -> pure mempty  -- Failed to open .gitignore
+        Right mfIgnores -> either (throwError . ParseError) pure mfIgnores
+
+updateIndexDir
+  :: (MonadIO m, MonadError GitError m, MonadState Index m, MonadReader Repo m)
+  => Path.RelDir -> m ()
+updateIndexDir targetDir = do
+    repoRoot <- repoFilePath <$> ask
+    parentIgnores <- mapM getIgnores $ parentDirPaths repoRoot
+    addDirContents repoRoot (foldr (<>) mempty parentIgnores) targetDir
```

Comment:
Why is this a `foldr`? Is it to do with the order of tree traversal?

Reply:
The parent dirs come out like `[/foo/bar/baz, /foo/bar, /foo]` but we want the .gitignores to have the opposite priority.